File: Source/Aladdin.CAPI.GOST/ASN1/GOST/EncryptedKey.cs
using System;

//	GOST28147EncryptedKey ::= SEQUENCE {
//		encryptedKey				OCTET STRING (SIZE (32 | 64)),
//		maskKey      [0] IMPLICIT	OCTET STRING (SIZE (32 | 64)) OPTIONAL,
//		macKey						OCTET STRING (SIZE (4))
//	}

namespace Aladdin.ASN1.GOST
{
	public class EncryptedKey : Sequence
	{
		// информация о структуре
		private static readonly ObjectInfo[] info = new ObjectInfo[] { 

			new ObjectInfo(new ObjectCreator<OctetString>().Factory(32, 64), Cast.N, Tag.Any		), 
			new ObjectInfo(new ObjectCreator<OctetString>().Factory(32, 64), Cast.O, Tag.Context(0)	), 
			new ObjectInfo(new ObjectCreator<OctetString>().Factory( 4,  4), Cast.N, Tag.Any		), 
		}; 
		// конструктор при раскодировании
		public EncryptedKey(IEncodable encodable) : base(encodable, info) {}

		// конструктор при закодировании
		public EncryptedKey(OctetString encryptedKey, OctetString maskKey, OctetString macKey) : 
			base(info, encryptedKey, maskKey, macKey) {}

		public OctetString	Encrypted	{ get { return (OctetString)this[0]; } } 
		public OctetString	MaskKey		{ get { return (OctetString)this[1]; } }
		public OctetString	MacKey		{ get { return (OctetString)this[2]; } }
	}
}

File: Source/Aladdin.CAPI.GOST.Java/src/aladdin/asn1/gost/EncryptedKey.java
package aladdin.asn1.gost;
import aladdin.asn1.*; 
import java.io.*; 

//	GOST28147EncryptedKey ::= SEQUENCE {
//		encryptedKey				OCTET STRING (SIZE (32 | 64)),
//		maskKey      [0] IMPLICIT	OCTET STRING (SIZE (32 | 64)) OPTIONAL,
//		macKey						OCTET STRING (SIZE (4))
//	}

public final class EncryptedKey extends Sequence<OctetString>
{
    private static final long serialVersionUID = 7690035948344465889L;
    
	// информация о структуре
	private static final ObjectInfo[] info = new ObjectInfo[] { 

		new ObjectInfo(new ObjectCreator(OctetString.class).factory(32, 64), Cast.N,	Tag.ANY			), 
		new ObjectInfo(new ObjectCreator(OctetString.class).factory(32, 64), Cast.O,	Tag.context(0)	), 
		new ObjectInfo(new ObjectCreator(OctetString.class).factory( 4,  4), Cast.N,	Tag.ANY			), 
	}; 
	// конструктор при раскодировании
	public EncryptedKey(IEncodable encodable) throws IOException { super(encodable, info); }
    
	// конструктор при закодировании
	public EncryptedKey(OctetString encryptedKey, OctetString maskKey, OctetString macKey)
	{
		super(info, encryptedKey, maskKey, macKey); 
	}
	public final OctetString encrypted() { return get(0); } 
	public final OctetString maskKey  () { return get(1); }
	public final OctetString macKey	  () { return get(2); }
}
